### ISSUE TITLE
Typography: Support `as` prop on Text component for composition

### ISSUE THREAD

**Issue by anuraghazra** [razorpay/blade#1214]
## The Problem

I was writing the BottomSheet docs, and did this:

```jsx
<Text>
  By default BottomSheet's SnapPoints are <Text weight="bold">[35%, 50%, 85%]</Text>
</Text>
```

We support nested Text elements but those elements are all `display: block`, resulting in the bold text to place at the new line: 

<img width="1124" alt="image" src="https://github.com/razorpay/blade/assets/35374649/493df3bc-2029-409a-9a8f-bbcc1cee34c8">

Arranging the text with flex doesn't seem very optimal

```jsx
<Box display="flex" gap="spacing.what????">
 <Text>By default BottomSheet's SnapPoints are</Text>
 <Text weight="bold">[35%, 50%, 85%]</Text> 
</Box>
```

Flex approach creates few issues: 

- Introduces **unnecessary** wrappers & not feasible when we need to modify multiple parts of the same inline text. 
- Developer has to **guess what the gap** between the two Text elements should be. ( This should be handled by the word spacing only not flex gap)
- **Word breaks won't work** properly, When the screen size gets smaller the text will split apart in the middle even if there is space for the text to flow **(this is a major issue)**. [Check this Loom Video](https://www.loom.com/share/5d0d6d1b017745ce8edea420bdcc957a) on how flex is behaving vs spans. 

## The Solution

In a traditional web dev approach, I will either do display: inline or use spans to solve this case.

Should we support `as="span"` or option to make the Text component render inline? ([Reshaped supports this](https://reshaped.so/content/docs/utilities/text))

```jsx
<Text as="span"> // or `isInline` or just `display="inline-block"`
  By default BottomSheet's SnapPoints are <Text weight="bold">[35%, 50%, 85%]</Text>
</Text>
```

We've done this correctly in Link component which has `display: inline-block`

<img width="1155" alt="image" src="https://github.com/razorpay/blade/assets/35374649/f419fd89-3666-4bac-bdf3-409ea657bdeb">

**Comment by divyanshu013** [comment]
Overall factors:
- Nesting (for all typography incl text, headings) - what all constraints should be there - should work close to html behavior (related #812, #1043, #951)
- as prop
- Inheritance of css, how should this work with props.
- Native has differences, how to keep things consistent (related #1177)

**Comment by saurabhdaware** [comment]
I think since we already support styledProps, we might as well expose `display` property for consumers. It makes sense to restrict it in something like Badge where things change visually, but Text I don't see any harm.

We have also received some request earlier #1164 to be able to hide something in mobile. Will make that usecase simpler as well and will not cause any issues that divyanshu mentioned.

**Comment by saurabhdaware** [comment]
Referring to this request https://razorpay.slack.com/archives/C01H13RTF8V/p1685435722633369?thread_ts=1685435697.924759&cid=C01H13RTF8V 

I think there are more usecases where we would want to render Text, Heading, Title has span. So what Anurag proposed makes sense and would solve more usecases.

**Comment by cseas** [comment]
Found another [use-case][1] for need of `display='inline'` on Text when followed by a Link component.

<img width="599" alt="Screenshot 2023-05-30 at 4 05 59 PM" src="https://github.com/razorpay/blade/assets/29686866/a911e805-19bf-4135-88ea-5829981e3847">

[1]: https://www.figma.com/file/ytpVeqLD5hQOb1WwrdjBsU/PP%2C-PB---Blade-adoption?type=design&node-id=166-113478&t=ckac4oqsI89sjhTI-0

**Comment by saurabhdaware** [comment]
I think we should make Badge display inline-block by default since it is not taking the full width.

And Link followed by Text can be implemented this way right now-
```
<Text>Some Text <Link>Link</Link></Text>
```

**Comment by divyanshu013** [comment]
Are we tracking the inline property changes elsewhere (for badge)? This issue is for typography, we might miss this otherwise

**Comment by saurabhdaware** [comment]
> Basically every typography comoponent will need to support as
>
> Originally [suggested by @chaitanyadeorukhkar](https://razorpay.slack.com/archives/G01B3LQ9H0W/p1685436676508289?thread_ts=1685436248.744909&cid=G01B3LQ9H0W)

To start, we can just support `as="span"` on all Typography components.


👍🏼 react this comment to support. 👀 react to oppose. Comment below to add more thoughts / ideas

**Comment by cseas** [comment]
👎🏼

- Creating a GitHub issue for every single attribute requirement adds a lot of delay and context sidetrack for product devs.
- We've already discussed that restricting the rendered element for Typography is an accessibility issue, which is increasing our tech debt everyday until Typography fully supports the `as` prop.

**Comment by kamleshchandnani** [comment]
how about just exposing `isInline` as a boolean prop? so we can internally render `span` on web and similar thing on RN respectively?

**Comment by saurabhdaware** [comment]
> Please add full support, this requirement has been discussed in multiple issues already.

By full support you mean supporting all values of `as`? Can you elaborate on what you're suggesting?

> Creating a GitHub issue for every single attribute requirement adds a lot of delay and context sidetrack for product devs. We've seen many examples of this: [[ref](https://razorpay.slack.com/archives/CMQ3RBHEU/p1684837897440279), [onScroll](https://razorpay.slack.com/archives/CMQ3RBHEU/p1685010483063419?thread_ts=1684837897.440279&cid=CMQ3RBHEU), [href](https://razorpay.slack.com/archives/CMQ3RBHEU/p1685441516277739?thread_ts=1685441325.576299&cid=CMQ3RBHEU), [display](https://razorpay.slack.com/archives/CMQ3RBHEU/p1685526177654729), [textAlign](https://razorpay.slack.com/archives/C01H13RTF8V/p1685433947717729?thread_ts=1685421988.808309&cid=C01H13RTF8V)]

Understandable about the context switching. We do unfortunately have to take different calls based on different props. While some props are obvious to support, some props may need discussion.

**Comment by cseas** [comment]
> By full support you mean supporting all values of as? 

Yes, delivering the ideal semantic html is the responsibility of the product devs. Right now we're unable to do that because we're blocked by Blade. There are many text-based semantic elements that could be required for specific situations, for example, [time][1]. Don't wanna open a GitHub issue for every single example of this that comes up in the future.

[1]: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/time

**Comment by saurabhdaware** [comment]
## Conclusion on this issue

We have discussed this within the team internally and **decided to add `as` prop with semantic HTML element values such as `span`, `abbr`, `figcaption`, `time`, etc** (Elements that don't change visuals of the component and define a certain semantic) (Person implementing can go through [HTML Elements Reference](https://developer.mozilla.org/en-US/docs/Web/HTML/Element) and see which values define semantics of text)

Adding a note below on why we avoid just opening up all values or why we have constraints in general since that came up several times in the multiple issues.

## Why we try to avoid opening up all values or Why the constraints in general

Sorry this is little offtopic to this issue but just wanted to clear some doubts. If you have something to discuss on this, I think we can discuss on separate issue and keep this issue limited to `<Text as="" />` discussions.

As you correctly mentioned [here](https://razorpay.slack.com/archives/CMQ3RBHEU/p1685984596259629?thread_ts=1685526177.654729&cid=CMQ3RBHEU)
> Devs aren't really trying to break Blade components on purpose so the possibility of a dev intentionally adding display and shipping a "broken"

Well the intent of constraints is not because we don't trust consumers to ship a non-broken code but rather to guide on what is supported from blade and what is not.

If we don't put these constraints, "what is possible from blade" can be confused with "what is supported from blade". So constraints are a way to tell - "Hey, maybe rethink of what you're trying to do. Maybe check with designers if this is expected UI? maybe see if its something that is custom for your App that can be standardised in your own repo? maybe its common across entire Razorpay so let design-system team know"

### Let me give you an example in context of this issue -

Here you started a thread regarding not having a Button as Text - [thread](https://razorpay.slack.com/archives/CMQ3RBHEU/p1685977877644389)

<img width="287" alt="image" src="https://github.com/razorpay/blade/assets/30949385/7e2ec33d-c1d5-48a2-8c31-a138a49eaa64">

Now imagine if we had supported `<Text as="button" onClick={} />` on Text component, a lot of devs might have thought that this is something that we "support" from blade and would've implemented it as is (and it's not dev's fault. They don't know what is supported from blade and what is not). 

But here, we do expect devs to question designers or ask designers to connect with blade's designers. And eventually as [Kamlesh correctly mentioned, the right UI here is probably using tertiary button](https://razorpay.slack.com/archives/CMQ3RBHEU/p1686026222473299?thread_ts=1685977877.644389&cid=CMQ3RBHEU).

The constraints exist to avoid such scenarios. Now obviously you might say what designers build should be consistent from get go and there shouldn't be this back-and-forth which I agree and that is what how things should be "ideally". Unfortunately figma doesn't allow us to restrict too much (if this was possible, then we could've achieved consistency on Chakra itself) so there will be few scenarios where designer might implement something that is restricted from blade (we can add reviews and try to solve for this but I am not sure if achieving "what designers give can be built as is without any questions" is possible ever or is even ideal. Even irrespective of Blade, we expect devs and design to collaborate on UI building. 

Do create a separate issue if you have concerns about overall philosophy. We can discuss over that issue and clear out unrelated conversations from other issues.

As of this issue, as discussed above we will be adding `as` prop with all semantic text element values.

**Comment by cseas** [comment]
Thanks @saurabhdaware! Supporting all semantic text elements sounds good for a long-term solution.
Can we have a similar conclusion on https://github.com/razorpay/blade/issues/812 as well?

**Comment by prabhasjoshi** [comment]
@saurabhdaware can we support this prop "as" for badge component also ?
